user/rcsget.man
.\" $Id: rcsget.man,v 11.8 2025/09/28 22:53:51 tom Exp $
.TH RCSGET 1 2025-09-28 "" "User commands"
.SH NAME
\fBrcsget\fR \-
RCS check-out utility
.SH SYNOPSIS
\fBrcsget\fI [options] [file-specifications]\fR
.SH DESCRIPTION
\fBRcsget\fR is a simple interface to \fIrcs\fR (\fBr\fRevision
\fBc\fRontrol \fBs\fRystem), which is used to extract files
from \fIrcs\fR archives.
.PP
\fBRcsget\fR checks out (extracts) files from an \fIrcs\fR
archive which have been checked in using \fBrcsput \fRor\fBcheckin\fR.
\fBRcsput\fR checks in files, retaining their modification
date as the check-in date.
\fB Rcsget\fR extracts a given version of one or more files,
and sets their modification date to match the check-in date.
.SH OPTIONS
\fBRcsget\fR is a directory-recursive interface to the \fBco\fR
utility which does special things with the file modification date.
Most of its options are passed to \fBcheckout\fR or \fBco\fR,
and then interpreted to determine the correct modification date.
Options special to \fBrcsget\fR are:
.TP 4n
.B \-a
permits \fBrcsget\fR to process directory names
beginning with ".".
.TP
.B \-d
directs \fBrcsget\fR to scan the \fIrcs\fR directories
to find the list of files which are checked-out.
Otherwise, \fBrcsget\fR
will select only those files which are already checked-out.
.TP
.B \-L
directs \fBrcsget\fR to follow symbolic links to subdirectories.
These are ignored otherwise.
.TP
.B \-n
no-op mode directs \fBrcsget\fR to show what files
would be checked out without actually checking them out.
.TP
.B \-R
(same as \-\fBd\fR)
causes \fBrcsget\fR to recur through the specified directories.
.TP
.BI \-T path
specify an alternate pathname for the \fBcheckout\fR utility.
This overrides the default "\fBcheckout\fR"
path.
.PP
\fBRcsget\fR also recognizes (and acts on) the "\-\fBq\fR" (quiet)
and "\-\fBx\fR" options of \fBcheckout\fR.
.SH OPERATIONS
To extract files using \fBrcsget\fR, you must provide one or more
names.
They may be either file or directory names.
.PP
\fBRcsget\fR resolves names
until it finds one of the form "RCS/\fIfile\fR,v" before invoking \fBco\fR.
If it finds a directory (other
than "RCS"), \fBrcsget\fR will recur to extract files at lower
levels.
.SH ENVIRONMENT
\fBRcsget\fR is written in C, and runs on POSIX systems.
.SH FILES
\fBRcsget\fR uses the following files:
.TP
\fBcheckout\fR
a utility which invokes \fBco\fR and then
modifies the file's modification date so that it matches the version-date
selected which was extracted via \fBco\fR.
.SH ANTICIPATED CHANGES
None.
.SH AUTHORS
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
checkin, checkout, rcsput, ded, co\ (1)


CHANGES
2025-09-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* package/debian/copyright: formatting

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	check-manpage

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	fixup-th-man

	* user/rcsget.man, user/vcs.man: check-manpage

	* user/rcsget.man, user/vcs.man: fixup-th-man

	* configure: regen

	* aclocal.m4, configure.in: use CF_PROG_INSTALL

2025-07-09  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: update to 2025-07-10

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: fix license

2025-01-17  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* configure.in: use CF_WITH_SYSTYPE early

2025-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/src/checkin.c, src/checkout/src/checkout.c,
	  src/checkup/src/checkup.c, src/checkout/src/rcsget.c,
	  src/copy/src/copy.c, src/link2rcs/src/link2rcs.c,
	  src/permit/src/permit.c, src/vcs/src/delete.c, src/vcs/src/unlock.c,
	  src/checkin/src/rcsput.c, src/baseline/src/baseline.c:
	gcc15-warnings

2024-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024/07/27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024/05/27

2024-04-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-01-22  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec: bump

	* configure: regen

	* bin/makefile.in:
	build-fix for Solaris: rename variable $(USER) to $(AS_USER)

	* package/debian/copyright, COPYING: bump

	* aclocal.m4, configure.in: use pkg-config to find td_lib

	* package/debian/changelog: bump

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* aclocal.m4: build-fixes for NetBSD

	* configure: regen

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/makefile.in, src/vcs/src/makefile.in,
	  src/baseline/test/makefile.in, src/baseline/makefile.in,
	  src/checkin/test/makefile.in, src/checkout/src/makefile.in,
	  src/checkout/test/makefile.in, src/checkup/src/makefile.in,
	  src/checkup/test/makefile.in, src/copy/src/makefile.in,
	  src/copy/test/makefile.in, src/link2rcs/src/makefile.in,
	  src/link2rcs/test/makefile.in, support/makefile.in:
	use VPATH

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/makefile.in, src/checkout/makefile.in,
	  src/checkup/makefile.in, src/copy/makefile.in,
	  src/link2rcs/makefile.in, src/permit/makefile.in, src/vcs/makefile.in:
	use VPATH

	* package/debian/compat: bump

	* src/baseline/src/makefile.in, src/checkin/src/makefile.in,
	  src/permit/src/makefile.in, src/permit/test/makefile.in,
	  src/vcs/test/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/checkup.man: fix typo

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/rcsput.man, user/rcsget.man, user/permit.man, user/link2rcs.man,
	  user/checkup.man, user/checkin.man, user/copy.man, user/checkout.man,
	  user/baseline.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

2019-12-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* src/checkin/src/rcsput.c, src/checkin/src/checkin.c:
	use DYN-argv lists to replace catarg()

	* src/vcs/src/vcs.c, src/permit/src/permit.c:
	use show_argv2(), to handle verb vs argv[0]

	* src/checkout/src/checkout.c, src/checkout/src/rcsget.c:
	use DYN-argv lists to replace catarg()


src/checkout/src/rcsget.c
/*
 * Title:	rcsget.c (rcs get-tree)
 * Author:	T.E.Dickey
 * Created:	19 Oct 1989
 * Modified:
 *		05 Dec 2019, use DYN-argv lists.
 *		04 Sep 2012, pass-thru -c option for 'checkout'.
 *		12 Nov 1994, pass-thru -f, -k options for 'co'.
 *		22 Sep 1993, gcc warnings
 *		06 Feb 1992, revised filename-parsing with 'rcsargpair()',
 *			     obsoleted "-x" option.
 *		11 Oct 1991, converted to ANSI
 *		25 Sep 1991, added options R and L. Ensure that RCS-directory
 *			     exists before trying to extract the file.
 *			     Make this show normal-trace when "-n -q" are set.
 *		20 Jun 1991, use 'shoarg()'
 *		06 Jun 1991, use "-x" option in local name-checking
 *		03 Jun 1991, pass-thru "-x" to 'checkout'
 *		03 Jun 1991, mods to compile on apollo sr10.3
 *		19 Apr 1990, added "-T" option so that 'checkout' isn't
 *			     hard-coded.
 *		18 Apr 1990, changed call on rcs2name/name2rcs to support "-x"
 *			     option in checkin/checkout
 *		16 Apr 1990, interpret "-q" (quiet) option in this program
 *		03 Nov 1989, additional correction: if file does not exist, it
 *			     is ok to ask 'checkout' to get it!  Added a hack
 *			     ("-?" option) to get checkout to show its options
 *			     in this usage-message.
 *		01 Nov 1989, walktree passes null pointer to stat-block if
 *			     no-access.
 *
 * Function:	Use 'checkout' to checkout one or more files from the
 *		RCS-directory which is located in the current working
 *		directory, and then, to set the modification date of the
 *		checked-out files according to the last delta date (rather
 *		than the current date, as RCS assumes).
 *
 * Options:	(see 'usage()')
 */

#define	STR_PTYPES
#include	<ptypes.h>
#include	<dyn_str.h>
#include	<rcsdefs.h>
#include	<sccsdefs.h>
#include	<errno.h>

MODULE_ID("$Id: rcsget.c,v 11.11 2025/01/07 00:52:02 tom Exp $")

#define	VERBOSE	if (!quiet) PRINTF

static char user_wd[BUFSIZ];	/* working-directory for scan_archive */
static const char *verb = "checkout";
static int a_opt;		/* all-directory scan */
static int R_opt;		/* recur/directory-mode */
static int L_opt;		/* follow links */
static int n_opt;		/* no-op mode */
static int quiet;		/* "-q" option */

static void
set_wd(const char *path)
{
    if (!n_opt)
	if (chdir(path) < 0)
	    failed(path);
}

static ARGV *co_dyns;

static void
Checkout(char *working, char *archive)
{
    ARGV *list;

    list = argv_init();
    argv_append(&list, verb);
    argv_merge(&list, co_dyns);
    argv_append(&list, working);
    argv_append(&list, archive);

    if (!quiet || n_opt) {
	show_argv(stdout, argv_values(list));
    }
    if (!n_opt) {
	if (executev(argv_values(list)) < 0)
	    failed(working);
    }
    argv_free(&list);
}

static int
an_archive(const char *name)
{
    size_t len_name = strlen(name);
    size_t len_type = strlen(RCS_SUFFIX);
    return (len_name > len_type
	    && !strcmp(name + len_name - len_type, RCS_SUFFIX));
}

/*
 * Test for directories that we don't try to scan
 */
static int
ignore_dir(char *path)
{
    if (!a_opt && *pathleaf(path) == '.'
	&& sameleaf(path, sccs_dir((char *) 0, path))) {
	if (!quiet)
	    PRINTF("...skip %s\n", path);
	return TRUE;
    }
    return FALSE;
}

static void
Ignore(const char *name, const char *why)
{
    VERBOSE("?? ignored: %s%s\n", name, why);
}

/*ARGSUSED*/
static int
WALK_FUNC(scan_archive)
{
    char tmp[BUFSIZ];

    (void) level;

    if (!strcmp(user_wd, path))	/* account for initial argument */
	return (readable);
    if (!isFILE(sp->st_mode)
	|| !an_archive(name)) {
	Ignore(name, " (not an archive)");
	return (-1);
    }
    if (!strcmp(vcs_file((char *) 0, strcpy(tmp, name), FALSE), name))
	return (readable);

    set_wd(user_wd);
    Checkout(rcs2name(name, FALSE), pathcat(tmp, rcs_dir(NULL, NULL), name));
    set_wd(path);
    return (readable);
}

static int
WALK_FUNC(scan_tree)
{
    char tmp[BUFSIZ];
    char *s = pathcat(tmp, path, name);
    Stat_t sb;

    if (RCS_DEBUG)
	PRINTF("++ %s%sscan (%s, %s, %s%d)\n",
	       R_opt ? "R " : "",
	       L_opt ? "L " : "",
	       path, name, (sp == NULL) ? "no-stat, " : "", level);

    if (!quiet || n_opt)
	track_wd(path);

    if (sp == NULL) {
	if (R_opt && (level > 0)) {
	    Ignore(name, " (no such file)");
	}
    } else if (isDIR(sp->st_mode)) {
	abspath(s);		/* get rid of "." and ".." names */
	if (ignore_dir(s))
	    readable = -1;
	else if (sameleaf(s, rcs_dir(NULL, NULL))) {
	    if (R_opt) {
		(void) walktree(strcpy(user_wd, path),
				name, scan_archive, "r", level);
	    }
	    readable = -1;
	} else {
#ifdef	S_IFLNK
	    if (!L_opt
		&& (lstat(s, &sb) < 0 || isLINK(sb.st_mode))) {
		Ignore(name, " (is a link)");
		readable = -1;
	    }
#endif
	}
    } else if (!isFILE(sp->st_mode)) {
	Ignore(name, RCS_DEBUG ? " (not a file)" : "");
	readable = -1;
    }
    return (readable);
}

static void
do_arg(const char *name)
{
#ifdef	S_IFLNK
    if (!L_opt) {
	Stat_t sb;
	if (lstat(name, &sb) >= 0 && isLINK(sb.st_mode)) {
	    Ignore(name, " (is a link)");
	    return;
	}
    }
#endif
    (void) walktree((char *) 0, name, scan_tree, "r", 0);
}

static void
usage(int option)
{
    static const char *tbl[] =
    {
	"usage: rcsget [options] files"
	,""
	,"Options include all CHECKOUT options, plus:"
	,"  -a       process all directories, including those beginning with \".\""
	,"  -d       directory-mode (scan based on archives, rather than working files"
	,"  -L       follow symbolic-links to subdirectories when -R is set"
	,"  -n       no-op (show what would be checked-out, but don't do it"
	,"  -q       quiet (also passed to \"checkout\")"
	,"  -R       recur (same as -d)"
	,"  -T TOOL  specify alternate tool to \"checkout\" to invoke per-file"
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    if (option == '?')
	(void) execute(verb, "-?");
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    int j;
    char *s, *t;

    track_wd((char *) 0);

    /* process options */
    for (j = 1; (j < argc) && (*(s = argv[j]) == '-'); j++) {
	t = s + strlen(s);
	if (strchr("cfklpqrcswj", s[1]) != NULL) {
	    argv_append(&co_dyns, s);
	    if (s[1] == 'q')
		quiet = TRUE;
	} else
	    while (s[1]) {
		switch (s[1]) {
		case 'a':
		    a_opt = TRUE;
		    break;
		case 'R':
		case 'd':
		    R_opt = TRUE;
		    break;
		case 'L':
		    L_opt = TRUE;
		    break;
		case 'n':
		    n_opt = TRUE;
		    break;
		case 'T':
		    verb = s + 2;
		    s = t;
		    break;
		default:
		    usage(s[1]);
		}
		s++;
	    }
    }

    /* process filenames */
    if (j < argc) {
	while (j < argc) {
	    char working[MAXPATHLEN];
	    char archive[MAXPATHLEN];
	    Stat_t sb;

	    j = rcsargpair(j, argc, argv);
	    if (rcs_working(working, &sb) < 0 && errno != EISDIR)
		failed(working);

	    if (isDIR(sb.st_mode)) {
		if (!ignore_dir(working))
		    do_arg(working);
	    } else {
		(void) rcs_archive(archive, (Stat_t *) 0);
		Checkout(working, archive);
	    }
	}
    } else
	do_arg(".");

    exit(SUCCESS);
    /*NOTREACHED */
}
